From 36ac4c4a094f682f716622c745e53742203df477 Mon Sep 17 00:00:00 2001
From: Brad Ho <brad_ho@phoenix.com>
Date: Fri, 11 Mar 2022 15:55:37 +0800
Subject: [PATCH] Improve trigger signal for add sel and oem sel

---
 src/sel_logger.cpp | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/src/sel_logger.cpp b/src/sel_logger.cpp
index 591aa9c..d0b41f1 100755
--- a/src/sel_logger.cpp
+++ b/src/sel_logger.cpp
@@ -46,6 +46,8 @@ using SELCreated =
     sdbusplus::xyz::openbmc_project::Logging::SEL::Error::Created;
 #endif
 
+std::shared_ptr<sdbusplus::asio::dbus_interface> ifaceAddSel;
+
 struct DBusInternalError final : public sdbusplus::exception_t
 {
     const char* name() const noexcept override
@@ -196,6 +198,28 @@ static void toHexStr(const std::vector<uint8_t>& data, std::string& hexStr)
     hexStr = stream.str();
 }
 
+static void triggerSignal(
+    const std::vector<uint8_t>& eventData,
+    const uint16_t& genID)
+{
+    sdbusplus::message::message msg;
+
+    msg = ifaceAddSel->new_signal("SelAdd");
+    msg.append(eventData, genID);
+    msg.signal_send();
+}
+
+static void triggerSignal(
+    const std::vector<uint8_t>& eventData,
+    const uint8_t& recordType)
+{
+    sdbusplus::message::message msg;
+
+    msg = ifaceAddSel->new_signal("OEMSelAdd");
+    msg.append(eventData, recordType);
+    msg.signal_send();
+}
+
 template <typename... T>
 static uint16_t selAddSystemRecord([[maybe_unused]] const std::string& message,
                                    const std::string& path,
@@ -211,6 +235,8 @@ static uint16_t selAddSystemRecord([[maybe_unused]] const std::string& message,
     std::string selDataStr;
     toHexStr(selData, selDataStr);
 
+    triggerSignal(selData, genId);
+
 #ifdef SEL_LOGGER_SEND_TO_LOGGING_SERVICE
     using namespace xyz::openbmc_project::Logging::SEL;
     auto entryID = report<SELCreated>(
@@ -243,6 +269,8 @@ static uint16_t selAddOemRecord([[maybe_unused]] const std::string& message,
     std::string selDataStr;
     toHexStr(selData, selDataStr);
 
+    triggerSignal(selData, recordType);
+
 #ifdef SEL_LOGGER_SEND_TO_LOGGING_SERVICE
     using namespace xyz::openbmc_project::Logging::SEL;
     auto entryID = report<SELCreated>(
@@ -271,7 +299,7 @@ int main(int, char*[])
     auto server = sdbusplus::asio::object_server(conn);
 
     // Add SEL Interface
-    std::shared_ptr<sdbusplus::asio::dbus_interface> ifaceAddSel =
+    ifaceAddSel =
         server.add_interface(ipmiSelPath, ipmiSelAddInterface);
 
     // Add a new SEL entry
