From bb7434007e40bf04a0d3c57931e14a6df66a1d20 Mon Sep 17 00:00:00 2001
From: Brad Ho <brad_ho@phoenix.com>
Date: Fri, 14 Jan 2022 15:55:49 +0800
Subject: [PATCH] Implement add sel to trigger dbus signal

---
 src/sel_logger.cpp | 43 ++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 42 insertions(+), 1 deletion(-)

diff --git a/src/sel_logger.cpp b/src/sel_logger.cpp
index 591aa9c..30e0daf 100755
--- a/src/sel_logger.cpp
+++ b/src/sel_logger.cpp
@@ -46,6 +46,7 @@ using SELCreated =
     sdbusplus::xyz::openbmc_project::Logging::SEL::Error::Created;
 #endif
 
+
 struct DBusInternalError final : public sdbusplus::exception_t
 {
     const char* name() const noexcept override
@@ -260,6 +261,40 @@ static uint16_t selAddOemRecord([[maybe_unused]] const std::string& message,
 #endif
 }
 
+static void triggerSignal(
+    std::shared_ptr<sdbusplus::asio::dbus_interface> interface,
+    const std::vector<uint8_t>& eventData,
+    const uint16_t& genID)
+{
+    if (eventData.size() > selEvtDataMaxSize)
+    {
+        throw std::invalid_argument("Event data too large");
+    }
+
+    sdbusplus::message::message msg;
+
+    msg = interface->new_signal("SelAdd");
+    msg.append(eventData, genID);
+    msg.signal_send();
+}
+
+static void triggerOEMSelSignal(
+    std::shared_ptr<sdbusplus::asio::dbus_interface> interface,
+    const std::vector<uint8_t>& eventData,
+    const uint8_t& recordType)
+{
+    if (eventData.size() > selOemDataMaxSize)
+    {
+        throw std::invalid_argument("Event data too large");
+    }
+
+    sdbusplus::message::message msg;
+
+    msg = interface->new_signal("OEMSelAdd");
+    msg.append(eventData, recordType);
+    msg.signal_send();
+}
+
 int main(int, char*[])
 {
     // setup connection to dbus
@@ -271,7 +306,7 @@ int main(int, char*[])
     auto server = sdbusplus::asio::object_server(conn);
 
     // Add SEL Interface
-    std::shared_ptr<sdbusplus::asio::dbus_interface> ifaceAddSel =
+    static std::shared_ptr<sdbusplus::asio::dbus_interface> ifaceAddSel =
         server.add_interface(ipmiSelPath, ipmiSelAddInterface);
 
     // Add a new SEL entry
@@ -279,6 +314,9 @@ int main(int, char*[])
         "IpmiSelAdd", [](const std::string& message, const std::string& path,
                          const std::vector<uint8_t>& selData,
                          const bool& assert, const uint16_t& genId) {
+
+            triggerSignal(ifaceAddSel, selData, genId);
+
             return selAddSystemRecord(message, path, selData, assert, genId);
         });
     // Add a new OEM SEL entry
@@ -286,6 +324,9 @@ int main(int, char*[])
         "IpmiSelAddOem",
         [](const std::string& message, const std::vector<uint8_t>& selData,
            const uint8_t& recordType) {
+
+            triggerOEMSelSignal(ifaceAddSel, selData, recordType);
+
             return selAddOemRecord(message, selData, recordType);
         });
 
