From 0590d807c1d0a2c457f89bf48ed6969b425a74ce Mon Sep 17 00:00:00 2001
From: Kian_Lee <Kian_Lee@phoenix.com>
Date: Sun, 5 Dec 2021 21:32:10 -0500
Subject: [PATCH] Setting_identify_ID_LED

---
 src/chassiscommands.cpp | 33 +++++++++++++++++++++++++--------
 1 file changed, 25 insertions(+), 8 deletions(-)

diff --git a/src/chassiscommands.cpp b/src/chassiscommands.cpp
index 94bd6a3..468c9c0 100644
--- a/src/chassiscommands.cpp
+++ b/src/chassiscommands.cpp
@@ -46,10 +46,8 @@ static constexpr const char* interruptButtonPath =
 
 const static constexpr char* idButtonProp = "ButtonPressed";
 
-const static constexpr char* ledService =
-    "xyz.openbmc_project.LED.GroupManager";
 const static constexpr char* ledIDOnObj =
-    "/xyz/openbmc_project/led/groups/enclosure_identify";
+    "/xyz/openbmc_project/led/physical/identify";
 const static constexpr char* ledIDBlinkObj =
     "/xyz/openbmc_project/led/groups/enclosure_identify_blink";
 const static constexpr char* ledInterface = "xyz.openbmc_project.Led.Group";
@@ -76,12 +74,30 @@ static ipmi::ServiceCache LEDService(ledInterface, ledIDBlinkObj);
 void enclosureIdentifyLed(const char* objName, bool isIdLedOn)
 {
     auto bus = getSdBus();
+    auto service = ipmi::getService(*bus,"xyz.openbmc_project.Led.Physical",ledIDOnObj);
+    constexpr auto LED_IS_ENABLE = "xyz.openbmc_project.Led.Physical.Action.On";
+    constexpr auto LED_IS_DISABLE = "xyz.openbmc_project.Led.Physical.Action.Off";
+    constexpr auto* ID_Led_Interface = "xyz.openbmc_project.Led.Physical";
+    constexpr auto* ID_Led_Property = "State";
+    std::string Set_LED_Status;
+
+    if (isIdLedOn)
+    {
+        Set_LED_Status = LED_IS_ENABLE;
+    }
+    else
+    {
+        Set_LED_Status = LED_IS_DISABLE;
+    }
 
     try
     {
-        std::string service = LEDService.getService(*bus);
-        setDbusProperty(*bus, service, objName, ledInterface, ledProp,
-                        isIdLedOn);
+        setDbusProperty(*bus,
+                        service,
+                        ledIDOnObj,
+                        ID_Led_Interface,
+                        ID_Led_Property,
+                        Set_LED_Status);
     }
     catch (const std::exception& e)
     {
@@ -187,12 +203,13 @@ ipmi::RspType<> ipmiChassisIdentify(std::optional<uint8_t> interval,
     uint8_t identifyInterval = interval.value_or(defaultIdentifyTimeOut);
     bool forceIdentify = force.value_or(0) & 0x01;
 
+    // make sure timer is stopped
     enclosureIdentifyLed(ledIDOnObj, false);
     identifyTimer->stop();
 
     if (identifyInterval || forceIdentify)
     {
-        enclosureIdentifyLed(ledIDBlinkObj, true);
+        enclosureIdentifyLed(ledIDOnObj, true);
         if (forceIdentify)
         {
             chassisIDState = ChassisIDState::indefinite;
@@ -207,7 +224,7 @@ ipmi::RspType<> ipmiChassisIdentify(std::optional<uint8_t> interval,
     else
     {
         chassisIDState = ChassisIDState::off;
-        enclosureIdentifyLed(ledIDBlinkObj, false);
+        enclosureIdentifyLed(ledIDOnObj, false);
     }
     return ipmi::responseSuccess();
 }
