diff --git a/apphandler.cpp b/apphandler.cpp
index d567fe38..22f6328f 100644
--- a/apphandler.cpp
+++ b/apphandler.cpp
@@ -691,9 +691,9 @@ auto ipmiAppGetSelfTestResults() -> ipmi::RspType<uint8_t, uint8_t>
     //      [2] 1b = Internal Use Area of BMC FRU corrupted.
     //      [1] 1b = controller update 'boot block' firmware corrupted.
     //      [0] 1b = controller operational firmware corrupted.
-    constexpr uint8_t notImplemented = 0x56;
+    constexpr uint8_t isImplemented = 0x55;
     constexpr uint8_t zero = 0;
-    return ipmi::responseSuccess(notImplemented, zero);
+    return ipmi::responseSuccess(isImplemented, zero);
 }
 
 static constexpr size_t uuidBinaryLength = 16;
