From 7439a9e3884d08099a19ad3a51ced66ee1a09f36 Mon Sep 17 00:00:00 2001
From: Adeli Yeh <Adeli_Yeh@phoenix.com>
Date: Tue, 12 Oct 2021 17:11:56 +0800
Subject: [PATCH] Add BMC Reset Cause for cold reset command

---
 globalhandler.cpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)
 mode change 100644 => 100755 globalhandler.cpp

diff --git a/globalhandler.cpp b/globalhandler.cpp
old mode 100644
new mode 100755
index c192cbb..aa88c50
--- a/globalhandler.cpp
+++ b/globalhandler.cpp
@@ -4,6 +4,7 @@
 #include <ipmid/utils.hpp>
 #include <phosphor-logging/elog-errors.hpp>
 #include <phosphor-logging/log.hpp>
+#include <fstream>
 #include <string>
 #include <xyz/openbmc_project/Common/error.hpp>
 #include <xyz/openbmc_project/State/BMC/server.hpp>
@@ -18,6 +19,20 @@ using BMC = sdbusplus::xyz::openbmc_project::State::server::BMC;
 
 void register_netfn_global_functions() __attribute__((constructor));
 
+void setBmcResetCause(int cause)
+{
+    std::ofstream fo("/var/log/bmc_reset_cause");
+
+    if (!fo) {
+        log<level::ERR>("Failed to open bmc_reset_cause fstream.");
+        return;
+    }
+
+    fo << cause << std::endl;
+
+    fo.close();
+}
+
 void resetBMC()
 {
     sdbusplus::bus::bus bus{ipmid_get_sd_bus_connection()};
@@ -38,6 +53,8 @@ void resetBMC()
  */
 ipmi::RspType<> ipmiGlobalReset()
 {
+    setBmcResetCause(1);
+
     try
     {
         resetBMC();
