diff --git a/apphandler.cpp b/apphandler.cpp
old mode 100644
new mode 100755
index 65bb159..6e930e5
--- a/apphandler.cpp
+++ b/apphandler.cpp
@@ -786,25 +786,40 @@ auto ipmiAppGetBtCapabilities()
                                  outputBufferSize, transactionTime, nrRetries);
 }
 
-auto ipmiAppGetSystemGuid() -> ipmi::RspType<std::array<uint8_t, 16>>
+ipmi::RspType<std::array<uint8_t, 16>> ipmiAppGetSystemGuid(ipmi::Context::ptr ctx)
 {
-    static constexpr auto bmcInterface =
-        "xyz.openbmc_project.Inventory.Item.Bmc";
     static constexpr auto uuidInterface = "xyz.openbmc_project.Common.UUID";
     static constexpr auto uuidProperty = "UUID";
 
-    ipmi::Value propValue;
+    std::string rfc4122Uuid;
+
     try
     {
-        // Get the Inventory object implementing BMC interface
-        auto busPtr = getSdBus();
-        auto objectInfo = ipmi::getDbusObject(*busPtr, bmcInterface);
+        ipmi::DbusObjectInfo objInfo;
+
+        boost::system::error_code ec =
+            ipmi::getDbusObject(ctx, uuidInterface, "/", "bios", objInfo);
+        if (ec)
+        {
+            log<level::ERR>("Failed GetSubTree for finding BMC UUID interface",
+                            entry("ERROR=%s", ec.message().c_str()),
+                            entry("INTERFACE=%s", uuidInterface));
+            return ipmi::responseUnspecifiedError();
+        }
+
+        ec = ipmi::getDbusProperty(ctx, objInfo.second, objInfo.first, uuidInterface,
+                                uuidProperty, rfc4122Uuid);
+        if (ec)
+        {
+            log<level::ERR>("Failed to get BMC UUID property",
+                            entry("ERROR=%s", ec.message().c_str()),
+                            entry("OBJECT=%s", objInfo.second.c_str()),
+                            entry("PATH=%s", objInfo.first.c_str()),
+                            entry("INTERFACE=%s", uuidInterface),
+                            entry("PROPERTY=%s", uuidProperty));
+            return ipmi::responseUnspecifiedError();
+        }
 
-        // Read UUID property value from bmcObject
-        // UUID is in RFC4122 format Ex: 61a39523-78f2-11e5-9862-e6402cfc3223
-        propValue =
-            ipmi::getDbusProperty(*busPtr, objectInfo.second, objectInfo.first,
-                                  uuidInterface, uuidProperty);
     }
     catch (const InternalFailure& e)
     {
@@ -814,7 +831,6 @@ auto ipmiAppGetSystemGuid() -> ipmi::RspType<std::array<uint8_t, 16>>
         return ipmi::responseUnspecifiedError();
     }
     std::array<uint8_t, 16> uuid;
-    std::string rfc4122Uuid = std::get<std::string>(propValue);
     try
     {
         // convert to IPMI format
