diff --git a/src/appcommands.cpp b/src/appcommands.cpp
index 10e3d13..074dd92 100644
--- a/src/appcommands.cpp
+++ b/src/appcommands.cpp
@@ -256,6 +256,63 @@ std::optional<MetaRevision> convertIntelVersion(std::string& s)
     return std::nullopt;
 }
 
+// Support both 2 solutions:
+// 1.Phoenix solution1  v0.01
+//   Major        0
+//   Minor        01
+// 2.Phoenix solution2  v0.02-1-g0ffc79513
+//   Major        0
+//   Minor        02
+//   buildNo      1
+//   MetaHash     0ffc79513
+std::optional<MetaRevision> convertPhoenixVersion(std::string& s)
+{
+    std::smatch results;
+    MetaRevision rev;
+    bool is_matched = false;
+
+    constexpr size_t matchedPtec1_0 = 3;
+    constexpr size_t matchedPtec1_1 = 4;
+    std::regex pattern1_0("v(\\d+?).(\\d+?)");
+    std::regex pattern1_1("v(\\d+?).(\\d+?)-(\\w+?)");  // with '-dirty' string
+
+    constexpr size_t matchedPtec2_0 = 5;
+    constexpr size_t matchedPtec2_1 = 6;
+    std::regex pattern2_0("v(\\d+?).(\\d+?)-(\\d+?)-g(\\w+?)");
+    std::regex pattern2_1("v(\\d+?).(\\d+?)-(\\d+?)-g(\\w+?)-(\\w+?)");  // with '-dirty' string
+
+    if ((std::regex_match(s, results, pattern1_0)) || (std::regex_match(s, results, pattern1_1)))
+    {
+        if ((results.size() == matchedPtec1_0) || (results.size() == matchedPtec1_1))
+        {
+            is_matched = true;
+        }
+    }
+    else if ((std::regex_match(s, results, pattern2_0)) || (std::regex_match(s, results, pattern2_1)))
+    {
+        if ((results.size() == matchedPtec2_0) || (results.size() == matchedPtec2_1))
+        {
+            is_matched = true;
+        }
+    }
+
+
+    if (is_matched == true)
+    {
+        rev.major = static_cast<uint8_t>(std::stoi(results[1]));
+        rev.minor = static_cast<uint8_t>(std::stoi(results[2]));
+        std::string versionString =
+            std::to_string(rev.major) + ":" + std::to_string(rev.minor);
+        phosphor::logging::log<phosphor::logging::level::INFO>(
+            "Get BMC version",
+            phosphor::logging::entry("VERSION=%s", versionString.c_str()));
+
+        return rev;
+    }
+
+    return std::nullopt;
+}
+
 RspType<uint8_t,  // Device ID
         uint8_t,  // Device Revision
         uint7_t,  // Firmware Revision Major
@@ -295,6 +352,10 @@ RspType<uint8_t,  // Device ID
         {
             std::optional<MetaRevision> rev =
                 convertIntelVersion(versionString);
+            if (rev == std::nullopt)
+            {
+                rev = convertPhoenixVersion(versionString);
+            }
             if (rev.has_value())
             {
                 MetaRevision revision = rev.value();
@@ -303,19 +364,11 @@ RspType<uint8_t,  // Device ID
                 revision.minor = (revision.minor > 99 ? 99 : revision.minor);
                 devId.fwMinor =
                     revision.minor % 10 + (revision.minor / 10) * 16;
-                try
-                {
-                    uint32_t hash = std::stoul(revision.metaHash, 0, 16);
-                    hash = bswap_32(hash);
-                    devId.aux = (revision.buildNo & 0xFF) + (hash & 0xFFFFFF00);
-                    fwVerInitialized = true;
-                }
-                catch (const std::exception& e)
-                {
-                    phosphor::logging::log<phosphor::logging::level::ERR>(
-                        "Failed to convert git hash",
-                        phosphor::logging::entry("ERROR=%s", e.what()));
-                }
+
+                // Phoenix own solution should not cover aux field.
+                // Set to 0 as default.
+                devId.aux = 0;
+                fwVerInitialized = true;
             }
         }
     }
