diff --git a/src/storagecommands.cpp b/src/storagecommands.cpp
index bd43a34..3d1eaae 100644
--- a/src/storagecommands.cpp
+++ b/src/storagecommands.cpp
@@ -1007,7 +1007,8 @@ ipmi::RspType<uint16_t, // Next Record ID
         return ipmi::responseUnspecifiedError();
     }
 
-    if (recordType == intel_oem::ipmi::sel::systemEvent)
+    if (recordType == intel_oem::ipmi::sel::systemEvent &&
+        targetEntryFields.size() > 4)
     {
         // Get the timestamp
         std::tm timeStruct = {};
@@ -1074,8 +1075,10 @@ ipmi::RspType<uint16_t, // Next Record ID
             systemEventType{timestamp, generatorID, evmRev, sensorType,
                             sensorNum, eventType, eventDir, eventData});
     }
-    else if (recordType >= intel_oem::ipmi::sel::oemTsEventFirst &&
-             recordType <= intel_oem::ipmi::sel::oemTsEventLast)
+    else if ((recordType == intel_oem::ipmi::sel::systemEvent &&
+              targetEntryFields.size() == 4) ||
+             (recordType >= intel_oem::ipmi::sel::oemTsEventFirst &&
+              recordType <= intel_oem::ipmi::sel::oemTsEventLast))
     {
         // Get the timestamp
         std::tm timeStruct = {};
@@ -1111,6 +1114,11 @@ ipmi::RspType<uint16_t, // Next Record ID
     return ipmi::responseUnspecifiedError();
 }
 
+static constexpr char const* ipmiSELService =    "xyz.openbmc_project.Logging.IPMI";
+static constexpr char const* ipmiSELPath = "/xyz/openbmc_project/Logging/IPMI";
+static constexpr char const* ipmiSELAddInterface = "xyz.openbmc_project.Logging.IPMI";
+static const std::string ipmiSELAddMessage = "SEL Entry";
+
 ipmi::RspType<uint16_t> ipmiStorageAddSELEntry(
     uint16_t recordID, uint8_t recordType, uint32_t timestamp,
     uint16_t generatorID, uint8_t evmRev, uint8_t sensorType, uint8_t sensorNum,
@@ -1127,7 +1135,33 @@ ipmi::RspType<uint16_t> ipmiStorageAddSELEntry(
         recordID, recordType, timestamp, generatorID, evmRev, sensorType,
         sensorNum, eventType, eventData1, eventData2, eventData3);
 
+    std::vector<uint8_t> eventData(9, 0xFF);
+    eventData[0] = generatorID;
+    eventData[1] = generatorID >> 8;
+    eventData[2] = evmRev;
+    eventData[3] = sensorType;
+    eventData[4] = sensorNum;
+    eventData[5] = eventType;
+    eventData[6] = eventData1;
+    eventData[7] = eventData2;
+    eventData[8] = eventData3;
+
+    std::shared_ptr<sdbusplus::asio::connection> bus = getSdBus();
+    sdbusplus::message::message writeSEL = bus->new_method_call(
+        ipmiSELService, ipmiSELPath, ipmiSELAddInterface, "IpmiSelAddOem");
+    writeSEL.append(ipmiSELAddMessage,  eventData, recordType);
+
     uint16_t responseID = 0xFFFF;
+    try
+    {
+        auto ret = bus->call(writeSEL);
+        ret.read(responseID);
+    }
+    catch (sdbusplus::exception_t& e)
+    {
+        std::cerr << "failed to call IpmiSelAdd method\n";
+    }
+
     return ipmi::responseSuccess(responseID);
 }
 
