diff --git a/ipmi-whitelist.conf b/ipmi-whitelist.conf
index 67bf9f2..d5f4a00 100644
--- a/ipmi-whitelist.conf
+++ b/ipmi-whitelist.conf
@@ -316,6 +316,11 @@
 0x32:0x8d:0xff7f   //<Intel OEM Platform>:<Get SSD Power>
 0x32:0x90:0x7f7f   //<Intel OEM Platform>:<Get-Set Serial Port Speed>
 0x32:0x91:0x7f7f   //<Intel OEM Platform>:<Clear CMOS>
+0x3c:0x00:0x7f7f   //<Phoenix OEM>:<ImgXfer Info>
+0x3c:0x01:0x7f7f   //<Phoenix OEM>:<ImgXfer Open>
+0x3c:0x02:0x7f7f   //<Phoenix OEM>:<ImgXfer SetOffset>
+0x3c:0x03:0x7f7f   //<Phoenix OEM>:<ImgXfer Close>
+0x3c:0x04:0x7f7f   //<Phoenix OEM>:<ImgXfer Delete>
 0x3e:0x02:0xff7f   //<Intel Managed Data Region>:<BMC Data Region Update Event Message>
 0x3e:0x20:0x7f7f   //<Intel Managed Data Region>:<BMC Region Status>
 0x3e:0x21:0x7f7f   //<Intel Managed Data Region>:<BMC Region Update Complete>
