diff --git a/src/sensorcommands.cpp b/src/sensorcommands.cpp
index 71df160..eb03814 100755
--- a/src/sensorcommands.cpp
+++ b/src/sensorcommands.cpp
@@ -1706,6 +1706,29 @@ ipmi::RspType<uint16_t,            // next record ID
     uint8_t* respStart = reinterpret_cast<uint8_t*>(hdr) + offset;
     std::vector<uint8_t> recordData(respStart, respStart + bytesToRead);
     uint16_t nextRecordId = lastRecord > recordID ? recordID + 1 : 0XFFFF;
+    switch(recordData[3])  // Checked SDR Type
+    {
+        case 1: // SDR Type 1
+	    if (bytesToRead >= 47)
+            {
+                recordData[47] |= 0xC0;
+            }
+            break;
+        case 2: // SDR Type 2
+	    if (bytesToRead >= 31)
+            {
+                recordData[31] |= 0xC0;
+	    }
+            break;
+	case 3: // SDR Type 3
+	    if (bytesToRead >= 16)
+            {
+                recordData[16] |= 0xC0;
+	    }
+            break;
+        default:
+            break;
+    }
     return ipmi::responseSuccess(nextRecordId, recordData);
 }
 /* end storage commands */
