From 79b983cc857644fd3b0ac40c99e9f9ed439de92e Mon Sep 17 00:00:00 2001
From: Adeli Yeh <Adeli_Yeh@phoenix.com>
Date: Thu, 25 Nov 2021 12:51:07 +0800
Subject: [PATCH] Add DCMI commands to white list

---
 ipmi-whitelist.conf | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/ipmi-whitelist.conf b/ipmi-whitelist.conf
index 9c28db0..fec3a92 100755
--- a/ipmi-whitelist.conf
+++ b/ipmi-whitelist.conf
@@ -148,6 +148,19 @@
 0x0c:0x20:0x7f7f   //<Transport>:<SOL Activating>
 0x0c:0x21:0x7f7f   //<Transport>:<Set SOL Configuration Parameters>
 0x0c:0x22:0xffff   //<Transport>:<Get SOL Configuration Parameters>
+0x2c:0x01:0xffff   //<Group Extension>:<DCMI>:<Get DCMI Capabilities Info>
+0x2c:0x02:0xffff   //<Group Extension>:<DCMI>:<Get Power Reading>
+0x2c:0x03:0xffff   //<Group Extension>:<DCMI>:<Get Power Limit>
+0x2c:0x04:0xffff   //<Group Extension>:<DCMI>:<Set Power Limit>
+0x2c:0x05:0xffff   //<Group Extension>:<DCMI>:<Activate/Deactivate Power Limit>
+0x2c:0x06:0xffff   //<Group Extension>:<DCMI>:<Get Asset Tag>
+0x2c:0x07:0xffff   //<Group Extension>:<DCMI>:<Get DCMI Sensor Info>
+0x2c:0x08:0xffff   //<Group Extension>:<DCMI>:<Set Asset Tag>
+0x2c:0x09:0xffff   //<Group Extension>:<DCMI>:<Get Management Controller Identifier String>
+0x2c:0x0a:0xffff   //<Group Extension>:<DCMI>:<Set Management Controller Identifier String>
+0x2c:0x10:0xffff   //<Group Extension>:<DCMI>:<Get Temp Readings>
+0x2c:0x12:0xffff   //<Group Extension>:<DCMI>:<Set DCMI Configuration Parameters>
+0x2c:0x13:0xffff   //<Group Extension>:<DCMI>:<Get DCMI Configuration Parameters>
 0x2c:0x1f:0xff7f   //<Group Extension>:<Get CPU PECI Package Config Data>
 0x2c:0x20:0xff7f   //<Group Extension>:<Get MDR Data Region Status>
 0x2c:0x21:0x7f7f   //<Group Extension>:<Get MDR Region Update Complete>
