From 38ac505a56be492ea1f7ec9b1c164ed38a45dc44 Mon Sep 17 00:00:00 2001
From: Kian_Lee <Kian_Lee@phoenix.com>
Date: Fri, 10 Dec 2021 02:17:30 -0500
Subject: [PATCH] Add OEM command 0x2e 0x03 to white list

---
 ipmi-whitelist.conf | 1 +
 1 file changed, 1 insertion(+)

diff --git a/ipmi-whitelist.conf b/ipmi-whitelist.conf
index fec3a92..bea49de 100755
--- a/ipmi-whitelist.conf
+++ b/ipmi-whitelist.conf
@@ -173,6 +173,7 @@
 0x2c:0x30:0x7f7f   //<Group Extension>:<Set TPM Configuration>
 0x2c:0x31:0x7f7f   //<Group Extension>:<TPM Configuration Update Complete>
 0x2c:0x37:0xff7f   //<Group Extension>:<Read PCIe Cable EEPROM Data>
+0x2e:0x03:0xffff   //<phoenix OEM>:<Set Netowrk>
 0x2e:0x04:0xffff // OEM-IPMI Commands to talk to Phosphor-pid-control
 0x2e:0x80:0xff7f   //<OpenBMC OEM>:<Blob Transfer>
 0x2e:0xC0:0x7f7f   //<Intel OEM>:<Enable/Disable Node Manager Policy Control>
