diff --git a/include/ssl_key_handler.hpp b/include/ssl_key_handler.hpp
index 39e83d7..59e0c1f 100644
--- a/include/ssl_key_handler.hpp
+++ b/include/ssl_key_handler.hpp
@@ -249,8 +249,16 @@ inline void generateSslCertificate(const std::string& filepath,
             // not before this moment
             X509_gmtime_adj(X509_get_notBefore(x509), 0);
             // Cert is valid for 10 years
+            // X509_gmtime_adj(X509_get_notAfter(x509),
+            //                 60L * 60L * 24L * 365L * 10L);
+            // Cert is valid for 398 days
+            // which is the maxium that The Certification Authority Browser Forum,
+            // also known as CA/Browser Forum, states as allowed.
+            // Subscriber Certificates issued on or after September 1, 2020 MUST have
+            // a Validity Period no greater than 398 days.
+            // see SC31 - Browser Alignment #195 https://github.com/cabforum/documents/pull/195
             X509_gmtime_adj(X509_get_notAfter(x509),
-                            60L * 60L * 24L * 365L * 10L);
+                            60L * 60L * 24L * 398L);
 
             // set the public key to the key we just generated
             X509_set_pubkey(x509, pPrivKey);
@@ -262,12 +270,18 @@ inline void generateSslCertificate(const std::string& filepath,
             X509_NAME_add_entry_by_txt(
                 name, "C", MBSTRING_ASC,
                 reinterpret_cast<const unsigned char*>("US"), -1, -1, 0);
+            X509_NAME_add_entry_by_txt(
+                name, "ST", MBSTRING_ASC,
+                reinterpret_cast<const unsigned char *>("Oregon"), -1, -1, 0);
+            X509_NAME_add_entry_by_txt(
+                name, "L", MBSTRING_ASC,
+                reinterpret_cast<const unsigned char *>("Beaverton"), -1, -1, 0);
             X509_NAME_add_entry_by_txt(
                 name, "O", MBSTRING_ASC,
-                reinterpret_cast<const unsigned char*>("OpenBMC"), -1, -1, 0);
+                reinterpret_cast<const unsigned char *>("Phoenix Technologies Ltd."), -1, -1, 0);
             X509_NAME_add_entry_by_txt(
                 name, "CN", MBSTRING_ASC,
-                reinterpret_cast<const unsigned char*>(cn.c_str()), -1, -1, 0);
+                reinterpret_cast<const unsigned char *>("ServerBMC"), -1, -1, 0);
             // set the CSR options
             X509_set_issuer_name(x509, name);
 
@@ -371,7 +385,7 @@ inline std::shared_ptr<boost::asio::ssl::context>
     // unauthenticated endpoints
     // mSslContext->set_verify_mode(boost::asio::ssl::verify_peer);
 
-    SSL_CTX_set_options(mSslContext->native_handle(), SSL_OP_NO_RENEGOTIATION);
+    SSL_CTX_set_options(mSslContext->native_handle(), SSL_OP_CIPHER_SERVER_PREFERENCE);
 
     BMCWEB_LOG_DEBUG << "Using default TrustStore location: " << trustStorePath;
     mSslContext->add_verify_path(trustStorePath);
@@ -385,27 +399,32 @@ inline std::shared_ptr<boost::asio::ssl::context>
     // There is a pull request to add this.  Once this is included in an asio
     // drop, use the right way
     // http://stackoverflow.com/questions/18929049/boost-asio-with-ecdsa-certificate-issue
-    if (SSL_CTX_set_ecdh_auto(mSslContext->native_handle(), 1) != 1)
+//    if (SSL_CTX_set_ecdh_auto(mSslContext->native_handle(), 1) != 1)
+//    {
+//        BMCWEB_LOG_ERROR << "Error setting tmp ecdh list\n";
+//    }
+
+    if (1 != SSL_CTX_set1_curves_list(mSslContext->native_handle(), "X448:X25519:secp521r1:secp384r1:prime256v1"))
     {
-        BMCWEB_LOG_ERROR << "Error setting tmp ecdh list\n";
+        BMCWEB_LOG_ERROR << "Error 2 setting tmp ecdh list\n";
     }
 
     std::string mozillaModern = "ECDHE-ECDSA-AES256-GCM-SHA384:"
                                 "ECDHE-RSA-AES256-GCM-SHA384:"
-                                "ECDHE-ECDSA-CHACHA20-POLY1305:"
-                                "ECDHE-RSA-CHACHA20-POLY1305:"
                                 "ECDHE-ECDSA-AES128-GCM-SHA256:"
-                                "ECDHE-RSA-AES128-GCM-SHA256:"
-                                "ECDHE-ECDSA-AES256-SHA384:"
-                                "ECDHE-RSA-AES256-SHA384:"
-                                "ECDHE-ECDSA-AES128-SHA256:"
-                                "ECDHE-RSA-AES128-SHA256";
+                                "ECDHE-RSA-AES128-GCM-SHA256";
 
     if (SSL_CTX_set_cipher_list(mSslContext->native_handle(),
                                 mozillaModern.c_str()) != 1)
     {
         BMCWEB_LOG_ERROR << "Error setting cipher list\n";
     }
+
+    if (SSL_CTX_set_ciphersuites(mSslContext->native_handle(),
+                "TLS_AES_256_GCM_SHA384:TLS_AES_128_GCM_SHA256") != 1)
+    {
+        BMCWEB_LOG_ERROR << "Error setting TLSv1.3 ciphersuites list\n";
+    }
     return mSslContext;
 }
 } // namespace ensuressl
