diff --git a/arch/arm/include/asm/arch-aspeed/ast-g5-intel.h b/arch/arm/include/asm/arch-aspeed/ast-g5-intel.h
index 64f4ed17bf..09aaeac50d 100755
--- a/arch/arm/include/asm/arch-aspeed/ast-g5-intel.h
+++ b/arch/arm/include/asm/arch-aspeed/ast-g5-intel.h
@@ -12,6 +12,37 @@
 
 #define AST_G5_INTEL 1
 
+/*EXTRST# mask for PFR platform
+* EXTRST# is used by PFR CPLD to keep BMC in
+* reset during firmware authentication, updates and recovery
+* this mask selects the modules to be reset along with BMC so that
+* Host function are intact.
+* (this is fis from Purley PFR )
+*/
+#define AST_WDT_RESET_MASK ( \
+	WDT_RESET_MASK_SPI  | 	 \
+	WDT_RESET_MASK_XDMA	|	 \
+	WDT_RESET_MASK_MCTP |	 \
+	WDT_RESET_MASK_ADC |     \
+    WDT_RESET_MASK_JTAG |    \
+	WDT_RESET_MASK_PECI |    \
+	WDT_RESET_MASK_CRT |     \
+	WDT_RESET_MASK_MIC |      \
+    WDT_RESET_MASK_SDIO |     \
+    WDT_RESET_MASK_HAC |      \
+    WDT_RESET_MASK_VIDEO |    \
+    WDT_RESET_MASK_HID11 |    \
+    WDT_RESET_MASK_USB11 |    \
+    WDT_RESET_MASK_USB20 |    \
+    WDT_RESET_MASK_GRAPHICS | \
+    WDT_RESET_MASK_MAC2 |     \
+    WDT_RESET_MASK_MAC1 |     \
+    WDT_RESET_MASK_I2C |      \
+    WDT_RESET_MASK_AHB |      \
+    WDT_RESET_MASK_COPROC |   \
+    WDT_RESET_MASK_ARM |      \
+    0)
+
 #ifndef __ASSEMBLY__
 int intel_force_firmware_jumper_enabled(void);
 int intel_failed_boot(void);
diff --git a/arch/arm/include/asm/arch-aspeed/regs-scu.h b/arch/arm/include/asm/arch-aspeed/regs-scu.h
index 1bdb1d8574..0a4fb6f773 100755
--- a/arch/arm/include/asm/arch-aspeed/regs-scu.h
+++ b/arch/arm/include/asm/arch-aspeed/regs-scu.h
@@ -144,6 +144,35 @@
 #define SCU_RESET_AHB			(0x1 << 1)
 #define SCU_RESET_SRAM_CTRL		(0x1 << 0)
 
+/* AST_WDT/EXTRST - 0x9C selection masks */
+#define WDT_RESET_MASK_MISC (1 << 25)    /* Misc. SOC controller (WDT, RTC, Timer, UART, SRAM.) */
+#define WDT_RESET_MASK_SPI (1 << 24)     /* SPI controller */
+#define WDT_RESET_MASK_XDMA (1 << 23)    /* X-DMA controller */
+#define WDT_RESET_MASK_MCTP (1 << 22)    /* MCTP controller */
+#define WDT_RESET_MASK_GPIO (1 << 21)    /* GPIO controller */
+#define WDT_RESET_MASK_ADC (1 << 20)     /* ADC controller */
+#define WDT_RESET_MASK_JTAG (1 << 19)    /* JTAG master controller */
+#define WDT_RESET_MASK_PECI (1 << 18)    /* PECI controller */
+#define WDT_RESET_MASK_PWM (1 << 17)     /* PWM controller */
+#define WDT_RESET_MASK_CRT (1 << 16)     /* CRT mode 2D engine */
+#define WDT_RESET_MASK_MIC (1 << 15)     /* MIC controller */
+#define WDT_RESET_MASK_SDIO (1 << 14)    /* SD/SDIO controller */
+#define WDT_RESET_MASK_LPC (1 << 13)     /* LPC controller */
+#define WDT_RESET_MASK_HAC (1 << 12)     /* HAC engine */
+#define WDT_RESET_MASK_VIDEO (1 << 11)   /* Video engine */
+#define WDT_RESET_MASK_HID11 (1 << 10)   /* USB1.1 HID/USB2.0 Host EHCI2 controller */
+#define WDT_RESET_MASK_USB11 (1 << 9)    /* USB1.1 Host controller */
+#define WDT_RESET_MASK_USB20 (1 << 8)    /* USB2.0 Host/Hub controller */
+#define WDT_RESET_MASK_GRAPHICS (1 << 7) /* Graphics CRT controller */
+#define WDT_RESET_MASK_MAC2 (1 << 6)     /* MAC#2 controller */
+#define WDT_RESET_MASK_MAC1 (1 << 5)     /* MAC#1 controller */
+#define WDT_RESET_MASK_I2C (1 << 4)      /* I2C controller */
+#define WDT_RESET_MASK_AHB (1 << 3)      /* AHB bridges */
+#define WDT_RESET_MASK_SDRAM (1 << 2)    /* SDRAM controller */
+#define WDT_RESET_MASK_COPROC (1 << 1)   /* Coprocessor */
+#define WDT_RESET_MASK_ARM (1 << 0)      /* ARM */
+#define WDT_RESET_MASK_ALL 0x03ffffff    /* all the bits above: 0-25*/
+
 /* AST_SCU_RESET2			0xD4 - Reset Control register set 2 */
 #define SCU_RESET_CRT3			(0x1 << 8)
 #define SCU_RESET_CRT2			(0x1 << 7)
diff --git a/board/aspeed/ast-g5/ast-g5-intel.c b/board/aspeed/ast-g5/ast-g5-intel.c
index b8aa5478d5..6741a94e5f 100755
--- a/board/aspeed/ast-g5/ast-g5-intel.c
+++ b/board/aspeed/ast-g5/ast-g5-intel.c
@@ -127,6 +127,14 @@ static const GPIOValue gpio_table[] = {
 /* HICRB Bits */
 #define HICRB_EN80HSGIO (1 << 13) /* Enable 80hSGIO */
 
+/* PCI Express Configuration Registers. */
+#define PCI_CONFIG_LOCK_KEY        0x0
+#define PCI_CONFIG_UNLOCK_KEY      0xa8
+#define PCI_PWR_CTRL_REG_OFFSET    0x7c
+#define PCI_L1_CONFIG_REG_OFFSET   0x68
+#define PCI_L1_ENABLE_VALUE        0xc81f0a
+#define PCI_L1_DISABLE_VALUE       0xc807fa
+
 #define SGPIO_CLK_DIV(N) ((N) << 16)
 #define SGPIO_BYTES(N) ((N) << 6)
 #define SGPIO_ENABLE 1
@@ -138,6 +146,7 @@ static const GPIOValue gpio_table[] = {
 #define WATCHDOG_RESET_BIT 0x8
 #define BOOT_FAILURE_LIMIT 0x3
 
+#define BRIGHTON_CITY_BOARD_ID 0
 #define COOPER_CITY_BOARD_ID 40
 
 static int get_boot_failures(void)
@@ -154,6 +163,18 @@ static void set_boot_failures(u32 count)
 	saveenv();
 }
 
+static inline void pcie_l1_disable()
+{
+	/* Unlock Power Management registers. */
+	writel(PCI_CONFIG_UNLOCK_KEY,
+		AST_PCIE_PLDA_BASE + PCI_PWR_CTRL_REG_OFFSET);
+	/* Disable PCIe L1. */
+	writel(PCI_L1_DISABLE_VALUE,
+		AST_PCIE_PLDA_BASE + PCI_L1_CONFIG_REG_OFFSET);
+	/* Relock the registers. */
+	writel(PCI_CONFIG_LOCK_KEY,
+		AST_PCIE_PLDA_BASE + PCI_PWR_CTRL_REG_OFFSET);
+}
 
 static void sgpio_init(void)
 {
@@ -667,6 +688,10 @@ void ast_g5_intel(void)
 	mailbox_init();
 	pwm_init();
 	gpio_init(gpio_table, ARRAY_SIZE(gpio_table));
+
+	/* PCH used in BrightonCity does not support PCIe L1.*/
+	pcie_l1_disable();
+
 	espi_init();
 	sgpio_init();
 	timer8_init();
