diff --git a/hook_add_entropy_buffer.c b/hook_add_entropy_buffer.c
index 0e1b828..638e223 100644
--- a/hook_add_entropy_buffer.c
+++ b/hook_add_entropy_buffer.c
@@ -1,7 +1,7 @@
 /* */
 #include "hook_add_entropy_buffer.h"
 
-size_t hook_add_entropy_buffer(void *buf, size_t size)
+size_t hook_add_entropy_buffer(int command, void *buf, size_t size, void* optional_buf, size_t optional_size);
 {
     return(0);
 }
