diff --git a/hook_add_entropy_buffer.h b/hook_add_entropy_buffer.h
index a729dbe..d0402d7 100644
--- a/hook_add_entropy_buffer.h
+++ b/hook_add_entropy_buffer.h
@@ -5,7 +5,14 @@
 
 #include <sys/types.h>
 
-size_t hook_add_entropy_buffer(void *buf, size_t size);
+#define CMD_NOOP          0
+#define CMD_CLR_BUFFER    1
+#define CMD_FILL_BUFFER   2
+#define CMD_SHRINK_BUFFER 3
+#define CMD_WHITEN_BUFFER 4
+
+
+size_t hook_add_entropy_buffer(int command, void *buf, size_t size, void *optional_buf, size_t optional_size);
 
 #endif /* RNGD_HOOK__H */
 
