From f51f0c1df5d7cffc21ffa732c64f1b8751ff3365 Mon Sep 17 00:00:00 2001
From: Adeli Yeh <Adeli_Yeh@phoenix.com>
Date: Fri, 19 Nov 2021 17:37:41 +0800
Subject: [PATCH] Remove SOL service check

---
 command/payload_cmds.cpp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/command/payload_cmds.cpp b/command/payload_cmds.cpp
index 4d67d1b..6b98bc3 100644
--- a/command/payload_cmds.cpp
+++ b/command/payload_cmds.cpp
@@ -45,7 +45,7 @@ std::vector<uint8_t> activatePayload(const std::vector<uint8_t>& inPayload,
         response->completionCode = IPMI_CC_INVALID_FIELD_REQUEST;
         return outPayload;
     }
-
+/*
     sol::Manager::get().updateSOLParameter(ipmi::convertCurrentChannelNum(
         ipmi::currentChNum, getInterfaceIndex()));
     if (!sol::Manager::get().enable)
@@ -53,7 +53,7 @@ std::vector<uint8_t> activatePayload(const std::vector<uint8_t>& inPayload,
         response->completionCode = IPMI_CC_PAYLOAD_TYPE_DISABLED;
         return outPayload;
     }
-
+*/
     // Only one instance of SOL is currently supported.
     if (request->payloadInstance != 1)
     {
