diff --git a/watch.cpp b/watch.cpp
old mode 100644
new mode 100755
index 90f75a7..ccdf594
--- a/watch.cpp
+++ b/watch.cpp
@@ -46,7 +46,7 @@ Watch::Watch(sd_event* loop, std::function<int(std::string&)> imageCallback) :
                                  std::strerror(error));
     }
 
-    wd = inotify_add_watch(fd, IMG_UPLOAD_DIR, IN_CLOSE_WRITE);
+    wd = inotify_add_watch(fd, IMG_UPLOAD_DIR, IN_CLOSE_WRITE | IN_MOVED_TO);
     if (-1 == wd)
     {
         auto error = errno;
@@ -97,7 +97,8 @@ int Watch::callback(sd_event_source* /* s */, int fd, uint32_t revents,
     while (offset < bytes)
     {
         auto event = reinterpret_cast<inotify_event*>(&buffer[offset]);
-        if ((event->mask & IN_CLOSE_WRITE) && !(event->mask & IN_ISDIR))
+        if ((event->mask & (IN_CLOSE_WRITE | IN_MOVED_TO)) &&
+            !(event->mask & IN_ISDIR))
         {
             auto tarballPath = std::string{IMG_UPLOAD_DIR} + '/' + event->name;
             auto rc = static_cast<Watch*>(userdata)->imageCallback(tarballPath);
